cpp/src/sampling/rw_traversals.hpp

```diff
@@ -0,0 +1,297 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Andrei Schaffer, aschaffer@nvidia.com
+//
+#pragma once
+
+#include <cugraph/experimental/graph.hpp>
+
+#include <utilities/graph_utils.cuh>
+
+#include <raft/device_atomics.cuh>
+#include <raft/handle.hpp>
+#include <raft/random/rng.cuh>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/for_each.h>
+
+#include <algorithm>
+#include <future>
+#include <thread>
+
+namespace cugraph {
+namespace experimental {
+
+namespace detail {
+
+template <typename T>
+using device_vec_t = rmm::device_uvector<T>;
+
+template <typename T>
+using device_v_it = typename device_vec_t<T>::iterator;
+
+template <typename value_t>
+value_t* raw_ptr(device_vec_t<value_t>& dv)
+{
+  return dv.data();
+}
+
+template <typename value_t>
+value_t const* raw_const_ptr(device_vec_t<value_t> const& dv)
+{
+  return dv.data();
+}
+
+template <typename value_t, typename index_t = size_t>
+struct device_const_vector_view {
+  device_const_vector_view(value_t const* d_buffer, index_t size) : d_buffer_(d_buffer), size_(size)
+  {
+  }
+
+  device_const_vector_view(device_const_vector_view const& other) = delete;
+  device_const_vector_view& operator=(device_const_vector_view const& other) = delete;
+
+  device_const_vector_view(device_const_vector_view&& other)
+  {
+    d_buffer_ = other.d_buffer_;
+    size_     = other.size_;
+  }
+  device_const_vector_view& operator=(device_const_vector_view&& other)
+  {
+    d_buffer_ = other.d_buffer_;
+    size_     = other.size_;
+
+    return *this;
+  }
+
+  value_t const* begin(void) const { return d_buffer_; }
+
+  value_t const* end() const { return d_buffer_ + size_; }
+
+  index_t size(void) const { return size_; }
+
+ private:
+  value_t const* d_buffer_{nullptr};
+  index_t size_;
+};
+
+template <typename value_t>
+value_t const* raw_const_ptr(device_const_vector_view<value_t>& dv)
+{
+  return dv.begin();
+}
+
+// classes abstracting the way the random walks path are generated:
+//
+
+// vertical traversal proxy:
+// a device vector of next vertices is generated for each path;
+// when a vertex is a sink the corresponding path doesn't advance anymore;
+//
+// smaller memory footprint;
+//
+struct vertical_traversal_t {
+  vertical_traversal_t(size_t num_paths, size_t max_depth)
+    : num_paths_(num_paths), max_depth_(max_depth)
+  {
+  }
+
+  template <typename graph_t,
+            typename random_walker_t,
+            typename index_t,
+            typename real_t,
+            typename seed_t>
+  void operator()(
+    graph_t const& graph,                // graph being traversed
+    random_walker_t const& rand_walker,  // random walker object for which traversal is driven
+    seed_t seed0,                        // initial seed value
+    device_vec_t<typename graph_t::vertex_type>& d_coalesced_v,  // crt coalesced vertex set
```

Comment:
crt implies current here?

Reply:
Yes.

Reply:
Hey @kaatish , can this be resolved so we can converge towards an approval? TIA.